src/notation/view/internal/dynamicpopupmodel.cpp

```diff
@@ -0,0 +1,225 @@
+        }
+        size_t symNameStart = startPos + startTag.length();
+        size_t endPos = text.find(endTag, symNameStart);
+        if (endPos == std::string::npos) {
+            break;
+        }
+        SymId id = SymNames::symIdByName(text.substr(symNameStart, endPos - symNameStart));
+        str += engravingFont->toString(id);
+        pos = endPos + endTag.length();
+    }
+
+    return str.toQString();
+}
+
+void DynamicPopupModel::init()
+{
+    AbstractElementPopupModel::init();
+
+    m_pages.clear();
+
+    if (!m_item) {
+        return;
+    }
+
+    IEngravingFontPtr engravingFont = m_item->score()->engravingFont();
+
+    for (const QList<DynamicPopupModel::PageItem>& page : DYN_POPUP_PAGES) {
+        QVariantList variantPage;
+        for (const DynamicPopupModel::PageItem& item : page) {
+            QVariantMap variantMap {
+                { "text", xmlTextToQString(item.text, engravingFont) },
+                { "width", item.width },
+                { "offset", item.offset },
+                { "type", item.itemType },
+            };
+            variantPage.append(variantMap);
+        }
+        m_pages.append(QVariant::fromValue(variantPage));
+    }
+
+    emit pagesChanged();
+}
+
+void DynamicPopupModel::addOrChangeDynamic(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
+        return;
+    }
+
+    beginCommand();
+    m_item->undoChangeProperty(Pid::TEXT, Dynamic::dynamicText(DYN_POPUP_PAGES[page][index].dynType));
+    m_item->undoChangeProperty(Pid::DYNAMIC_TYPE, DYN_POPUP_PAGES[page][index].dynType);
+    endCommand();
+
+    if (m_item->flag(ElementFlag::IS_PREVIEW)) {
+        m_item->setFlag(ElementFlag::IS_PREVIEW, false);
+    }
+
+    INotationInteractionPtr interaction = currentNotation()->interaction();
+
+    interaction->flipHairpinsType(toDynamic(m_item));
+
+    // Hide the bounding box which appears when called using Ctrl+D shortcut
+    if (interaction->isTextEditingStarted()) {
+        interaction->endEditText();
+        interaction->startEditGrip(m_item, Grip::DRAG);
+    }
+
+    updateNotation();
+}
+
+void DynamicPopupModel::addHairpinToDynamic(ItemType itemType)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
+        return;
+    }
+
+    beginCommand();
+
+    Hairpin* pin = Factory::createHairpin(m_item->score()->dummy()->segment());
+    if (itemType == ItemType::Crescendo) {
+        pin->setHairpinType(HairpinType::CRESC_HAIRPIN);
+    } else if (itemType == ItemType::Decrescendo) {
+        pin->setHairpinType(HairpinType::DECRESC_HAIRPIN);
+    }
+
+    m_item->score()->addHairpinToDynamic(pin, toDynamic(m_item));
+
+    endCommand();
+    updateNotation();
+}
+
+void DynamicPopupModel::showPreview(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic() || !m_item->flag(ElementFlag::IS_PREVIEW)) {
```

Comment:
Connected to what I said earlier about setting the preview flag here instead of in `engraving`: now I see why you did it in `engraving`, namely because you only want the preview functionality for newly created dynamics. 

I'd still propose to set the flag here instead of in `engraving`, and to detect whether it's a newly created dynamic, check whether its text is empty.